### ISSUE TITLE
add basic support for `dyn Trait` and `impl Trait`

### ISSUE THREAD

**Issue by nikomatsakis** [rust-lang/chalk#218]
**Status:** @KiChjang is working on the first steps.

---

The goal here is to extend chalk with basic support for `dyn Trait` and (opaque) `impl Trait` types. This involves a few steps:

- [x] extend chalk-parser to support `dyn` and `impl` syntax, along with the ast
- [x] extending `chalk-ir` to be able to represent such types
- [x] extending `chalk-solve` so that when we have to prove something like `dyn Foo: Foo`, we judge that to be true (similarly `impl Foo: Foo`) -- blocked on #216 
- [ ] adding tests for cases like
    - [ ] unifying `dyn Foo` with a struct type (error)
    - [ ] unifying `impl Foo` with a struct type (error)
    - [x] checking that `dyn Foo: Foo` (without any impl)
    - [x] checking that `dyn Foo: Bar` does *not* hold
    - [x] checking that `impl Foo: Foo`  (without any impl)
    - [x] checking that `impl Foo: Bar` does *not* hold
        - hmm, we're going to have to figure out "auto trait leakage"

This is judged to be a "medium" complex project, suitable for a dedicated beginner. After this work is done, the next step is https://github.com/rust-lang/chalk/issues/203.

**Comment by nikomatsakis** [comment]
## Mentoring instructions for extending chalk-parser

To be written

**Comment by nikomatsakis** [comment]
## Mentoring instructions for extending chalk-ir

Extending chalk-ir should be *relatively* straightforward, though there are some interesting raised by it. I think we want to extend the `Ty` type:

https://github.com/rust-lang/chalk/blob/6fd1808896bd5c6ec7cdfcedaa96b6939f7881d5/chalk-ir/src/lib.rs#L217

to include two new variants, `Dyn` and `Opaque`. These probably look like:

```rust
enum Ty {
    ...
    Dyn(Vec<QuantifiedWhereClause>), // dyn Trait
    Opaque(Vec<QuantifiedWhereClause>), // impl Trait
}
```

These are the counterpart to the existing `ForAll` type:

https://github.com/rust-lang/chalk/blob/6fd1808896bd5c6ec7cdfcedaa96b6939f7881d5/chalk-ir/src/lib.rs#L244

we then have to adjust the lowering code to lower the AST to them, which would be done "sort of vaguely" [like the existing code for `ForAll`](https://github.com/rust-lang/chalk/blob/6fd1808896bd5c6ec7cdfcedaa96b6939f7881d5/src/lowering.rs#L915-L930).

Note that these types introduce a level of binder -- this is so that you can represent the (unknown) `Self` type with a debruijn variable. i.e., something like `dyn Foo<Bar>` would wind up represented as

```
Dyn(^0: Foo<Bar>`)
```

where `^0` represents "debruijn variable with depth 0.

**Comment by nikomatsakis** [comment]
cc @scalexm -- I forget, how much of this did you do in https://github.com/rust-lang/chalk/pull/200 ?

**Comment by nikomatsakis** [comment]
(Also, the mentoring instructions here are fairly sketchy, consider it as future work)

**Comment by KiChjang** [comment]
Has anyone started working on this? If not, I'd like to try and take a stab at this.

**Comment by KiChjang** [comment]
I've opened #226 to further discuss the issues regarding the implementation on the chalk solver.

**Comment by nikomatsakis** [comment]
OK, @KiChjang, so what's missing from your PR is that we need to extend the "clause" lowering code. The basic rule that we want to create is that whenever the "self-type" is `dyn Foo` or `impl Foo`, we can add a rule that `dyn Foo: Foo` and `impl Foo: Foo`.

This would be done in the "clauses" code -- which is the code that takes a goal we are trying to prove and produces the "program clauses" that could be used to do so. Most of this work is done in this function right now:

https://github.com/rust-lang/chalk/blob/df09cc603c0cff9ed95e5554055d483fdd756fbc/chalk-solve/src/clauses.rs#L131-L140

and in particular I think the first thing we want to do is to extend this case, where we are trying to prove case like `Implemented(P0: SomeTrait<P1..Pn>)`. 

https://github.com/rust-lang/chalk/blob/df09cc603c0cff9ed95e5554055d483fdd756fbc/chalk-solve/src/clauses.rs#L142

What I think we want to do is to add a new case, right around [this "FIXME" line at the end](https://github.com/rust-lang/chalk/blob/df09cc603c0cff9ed95e5554055d483fdd756fbc/chalk-solve/src/clauses.rs#L170). 

In particular, we want to look at the *self type* and check whether it is a "dyn" or "impl" trait type. If you rebase this PR on master, you'll find a convenient helper on the `TraitRef` type caller `self_type_parameter` that you can use for this. The idea would be basically be to match the self-type and -- if it is a `dyn` or `impl` type -- then we know we are solving a goal like `dyn <Predicates>: ...`. 

These `<Predicates>` are actually a lambda like `exists<Self> { Predicates }`. So what we want to do is to replace the `Self` with the `dyn Type` -- so that if we had a type like `dyn Iterator`, we would get predicates like `dyn Iterator: Iterator`. We can then just push these into [the `clauses` vector](https://github.com/rust-lang/chalk/blob/df09cc603c0cff9ed95e5554055d483fdd756fbc/chalk-solve/src/clauses.rs#L139).

This isn't as targeted as it could be. For example, if we're being asked for program clauses to solve `Implemented(dyn Foo: Bar)`, we will wind up returning a program clause `Implemented(dyn Foo: Foo)`, which isn't really relevant. But that's ok, it'll get screened out later, and we can tweak how that works later.

**Comment by nikomatsakis** [comment]
The next step here is https://github.com/rust-lang/chalk/issues/275

**Comment by nikomatsakis** [comment]
So I think this is basically done. There are some caveats worth keeping in mind. For example, chalk assumes presently that the binders in a dyn/impl Trait are sorted into a canonical ordering. In any case, I'm going to close this, thanks @KiChjang ❤️
